docs/remote-attestation.md

```diff
@@ -0,0 +1,147 @@
+that is expected to run.
+
+One of the main components used in the remote attestation process is an
+**Attestation report**, which is a data structure signed by the TEE platform and
+containing information identifying the code that is running inside the TEE. This
+report can then be checked and signed by the **TEE Provider** (e.g. AMD or
+Intel), which results in evidence that the code is running on a genuine TEE
+platform.
+
+## Overview
+
+Remote Attestation protocol is implemented as a simplification of the
+[Enclave Key Exchange Protocol (EKEP)](https://asylo.dev/docs/concepts/ekep.html)
+handshake protocol.
+
+The Remote Attestation protocol works on top of the [gRPC](https://grpc.io/)
+protocol and consists of 2 stages:
+
+1. Handshake
+   - Server proves its identity by sending a cryptographically signed
+     Attestation Report
+   - Both Client and Server establish an encrypted connection with keys that are
+     bound to the TEE platform's firmware
+1. Data Exchange
+   - The keys generated during the Handshake are used to encrypt data sent
+     between Client and Server
+
+Prior to the _Handshake_ both Client and Server generate an individual pair of
+[ECDSA-P256](https://datatracker.ietf.org/doc/html/rfc6979) keys called
+**Signing Keys**. These keys are persistent between connections (but regenerated
+at startup) and are used to sign **Transcripts**: hashes of previous handshake
+messages that are sent in each new handshake message to prevent
+[Replay Attacks](https://en.wikipedia.org/wiki/Replay_attack). Each handshake
+message also includes a random string so that transcripts always contain new
+random values.
+
+Also for each individual connection both Client and Server generate a pair of
+[X25519](https://datatracker.ietf.org/doc/html/rfc7748) Diffie-Hellman keys
+called **Ephemeral Keys**. These keys are used to establish a shared secret
+between Client and Server.
+
+The shared secret is used to generate 2 shared
+[AES-256-GCM](https://datatracker.ietf.org/doc/html/rfc5288) **Session Keys**
+that are used during the _Data Exchange_ stage:
+
+- Server Session Key
+  - Which is used to encrypt messages sent by the Server
+- Client Session Key
+  - Which is used to encrypt messages sent by the Client
+
+## Workflow
+
+The workflow of the Remote Attestation protocol involves 3 interacting entities:
+
+1. Server
+   - Server application that processes requests from Clients
+   - Runs on a TEE Platform
+1. Client
+   - Client application that connects to the Server
+1. TEE Platform
+   - Firmware that provides a TEE support (i.e. Intel SGX or AMD-SEV-SNP capable
+     CPU)
+   - Possesses a firmware key
+1. TEE Provider
+   - TEE platform provider that can confirm authenticity of the TEE Platform
+     firmware keys using its root key
+   - Client must have the TEE Provider's root public key
+   - It’s important to note that the Provider is an external server (i.e.
```

Comment:
It is not always required to connect to an external server to validate the attestation report. Only one of the SGX attestation models requires this as far as I remember.